### PR TITLE
Added support for array index and corresponding unit test

### PR DESCRIPTION
**Pull request by jsboige**
That was trivial and a need on my side.

### CODE REVIEW

File: ExpressionUtils/Evaluating/ExpressionInterpreter.cs [hunk 1]
```diff
@@ -118,6 +118,8 @@ class ExpressionInterpretationVisitor : ExpressionResultVisitor<object> {
 							return left << right;
 						case ExpressionType.RightShift:
 							return left >> right;
+                        case ExpressionType.ArrayIndex:
+                            return ((Array) left).GetValue(right);
```
Comment (saraedum) [inline ExpressionUtils/Evaluating/ExpressionInterpreter.cs]: Since left is `dynamic`, is that cast actually necessary?

Comment (jsboige) [reply ExpressionUtils/Evaluating/ExpressionInterpreter.cs]: Alright I removed the cast

**Comment by saraedum** [comment]
Thanks for your contribution. I'll try to push a new release to nuget soon.

**Comment by saraedum** [comment]
I just released a new version. It would be great if you could check that your fixes works now :)

**Comment by jsboige** [comment]
Thanks, I was able to test that the new package includes arrays and I removed the temporary dependency on my build.

By the way, your tool was invaluable in supercharging what I think is a cool project: linq to z3.
This is [my tree](https://github.com/MyIntelligenceAgency/Z3.LinqBinding/tree/EPFdevelopment) with your last addition of binary array indexes used in the meal planner, which is a draft project for my students.

However, the partial evaluator and interpretor are used in several places all over the theorem engine to maximise the types of lambdas that can be ultimately translated into Z3 constraints to be solved or optimized. 
My upgraded version of Sudoku and MissionariesAndCannibals are probably good entry points to the flexibility that your library brought to the original prototype.

Would you mind having a look and tell me whether you think this is the correct approach to integrating your tool and if you can think of other magic it could add?

**Comment by saraedum** [comment]
That's a cool project. I wasn't aware of Z3 being that open by now.

Sure, I'll have a look at the integration. @Khazuar might also be curious to see how this is being used now.

**Comment by jsboige** [comment]
Thanks for your comments.

I believe I have addressed your first suggestion about scoping as can be seen [in the Theorem class](https://github.com/MyIntelligenceAgency/Z3.LinqBinding/blob/EPFdevelopment/src/Z3.LinqBinding/Theorem.cs#L413) and illustrated in [my upgraded version of Sudoku](https://github.com/MyIntelligenceAgency/Z3.LinqBinding/blob/EPFdevelopment/src/Z3.LinqBinding.Sudoku/SudokuAsArray.cs#L73). I found out about your library when I was specifically trying to solve that issue, and I'm very glad I did.
The result is much less verbose than the [original Sudoku example](https://github.com/MyIntelligenceAgency/Z3.LinqBinding/blob/EPFdevelopment/src/Z3.LinqBinding.Sudoku/SudokuTable.cs), for which the [tedious manipulation of Expressions](https://github.com/MyIntelligenceAgency/Z3.LinqBinding/blob/EPFdevelopment/src/Z3.LinqBinding.Sudoku/SudokuTheorem.cs) kind of takes away the original interest for the provider.

Now I'm not sure to understand your second suggestion about inlining. I think [this](https://github.com/MyIntelligenceAgency/Z3.LinqBinding/blob/EPFdevelopment/src/Z3.LinqBinding/Theorem.cs#L971) is the place where I wasn't so sure how to best further integrate your lib to handle mixed Member expressions. I only took care of a few cases with collections array indices,  indexed properties and Enumerable<T> functions. I wouldn't be surprise if you have got better suggestions to cover a whole range of LINQ uses here.
Also, what is the provider you originally wrote that lib for?

**Pull request merged**